librz/type/path.c

```diff
@@ -36,74 +36,199 @@ RZ_API void rz_type_path_free(RZ_NULLABLE RzTypePath *tpath) {
+		}
+
+		++*i; // skip '['
+		size_t tok_beg = *i;
+		for (; isdigit(path[*i]); ++*i)
+			;
+
+		if (path[*i] != ']') {
+			RZ_LOG_ERROR("Expected ']' got '%c'\n", path[*i]);
+			*offset = -1;
+			return NULL;
+		}
+
+		char *idx_str = rz_str_ndup(&path[tok_beg], *i - tok_beg);
+		size_t idx = rz_num_math(NULL, idx_str);
+		free(idx_str);
+
+		cur_dim_off /= typd->array.count;
+		*offset += cur_dim_off * idx;
+		typd = typd->array.type;
+
+		// DONT check for "idx < array size"
+		// because guy may be exploiting OOB or something else.
+		// so just continue.
+
+		++*i; // skip ']'
+	}
+
+	return typd;
+}
+
+/**
+ * \brief Parses the dot-separated token.
+ * \param typedb
+ * \param parent Previous token type.
+ * \param path String where struct path is stored.
+ * \param i Index in path pointing to the dot '.'.
+ * \param offset Value to place resulting offset on success. Set to -1 on error.
+ */
+static RzType *path_walker_parse_dot(const RzTypeDB *typedb, RZ_NONNULL RzType *parent, RZ_NONNULL const char *path, RZ_NONNULL RZ_INOUT size_t *i, RZ_NONNULL RZ_OUT st64 *offset) {
+	rz_return_val_if_fail(typedb && path, (*offset = -1, NULL));
+
+	++*i; // skip '.'
+	size_t tok_beg = *i;
+	for (; isalnum(path[*i]); ++*i)
+		;
+
+	char *tok = rz_str_ndup(&path[tok_beg], *i - tok_beg);
+
+	RzBaseType *parent_btype = rz_type_get_base_type(typedb, parent);
+	if (!parent_btype) {
+		RZ_LOG_ERROR("Could not found btype for parent '%s'\n", parent->identifier.name);
+		*offset = -1;
+		free(tok);
+		return NULL;
+	}
+
+	RzTypeStructMember *memb_it = NULL;
+	RzType *cur_type = NULL;
+	size_t cur_offset = -1;
+	rz_vector_foreach(&parent_btype->struct_data.members, memb_it) {
+		if (!strcmp(memb_it->name, tok)) {
+			cur_type = memb_it->type;
+			cur_offset = memb_it->offset; // in bytes
+			cur_offset *= 8; // in bits
+			break;
+		}
+	}
+
+	if (!cur_type) {
+		RZ_LOG_ERROR("Invalid member '%s' for parent type\n", tok);
+		*offset = -1;
+		free(tok);
+		return NULL;
+	}
+
+	free(tok);
+
+	*offset += cur_offset;
+	if (path[*i] == '[') {
+		parent = cur_type;
+		if (parent->kind != RZ_TYPE_KIND_ARRAY) {
+			RZ_LOG_ERROR("Expected array, got another type\n");
+			*offset = -1;
+			return NULL;
+		}
+
+		if (!path_walker_parse_bracket(typedb, parent, path, i, offset)) {
+			return NULL;
+		}
+
+		// unwrap array
+		while (parent->kind == RZ_TYPE_KIND_ARRAY) {
+			parent = parent->array.type;
+		}
+		return parent;
+	} else if (path[*i] == '.' || path[*i] == '\0') {
+		return cur_type;
+	}
+	RZ_LOG_ERROR("Unexpected character '%c' at position %lu\n", path[*i], *i);
+	*offset = -1;
+	return NULL;
+}
+
+/**
+ * \brief Walks through the given struct path and returns offset from base.
+ * \param typedb
+ * \param path Path to walk through (like "Struct.arr[30].member")
+ *
+ * \brief Splits struct path on tokens and iterates over them.
+ * \brief Calls `path_walker_parse_dot` on each.
+ */
+static st64 path_walker(const RzTypeDB *typedb, RZ_NONNULL const char *path) {
```

Comment:
```suggestion
static st64 path_walker(RZ_NONNULL const RzTypeDB *typedb, RZ_NONNULL const char *path) {
```

Reply:
fixed

Reply:
About the `tok_beg`. Here is how I will save the allocation.

The `rz_num_math` function internally determines the base of passed number. So using it allows to specify such paths: `"Struct.arr[0x20]"` which is useful from user's perspective. And it also frees me from determining the base myself.